### PR TITLE
Interactivity API: router with region-based client-side navigation

### PR DESCRIPTION
**Pull request by luisherranz**
## What?

This PR adds support for region-based client-side navigation.

It works with a new directive called `data-wp-navigation-id` which must be unique. Once those ids are in place, you can use `navigate` (exported for now in `@wordpress/interactivity`) to navigate to another page. Only the regions marked with `data-wp-navigation-id` will be replaced.

## Why?

Because with the Interactivity API, we can replace regions of the DOM with server-generated content while maintaining the UI state.

Use cases include:

- Client-side pagination of the Query block.
- Client-side submission of the Comments Form block.

## How?

Based on @darerodz's Woo PR: https://github.com/woocommerce/woocommerce-blocks/pull/10200

Apart from that, I added new options to `navigate` and `prefetch`:

- `force`: when `true`, it bypasses the client page cache.
- `html`: when it's a string, it doesn't fetch the URL and uses that HTML instead
- `replace`: when it's `true`, it uses `history.replaceState` instead of `history.pushState`.

## Tasks

- [x] Add changelog
- [x] Add e2e tests

We decided to leave these tasks for a following PR:
- Add JSDocs
- Add documentation

### CODE REVIEW

**Comment by ajgagnon** [comment]
This is great! This directive alone is so powerful.  🙌 

Do we need to handle loading states for slower connections or lower powered hosting?  

Livewire does this pretty well I think.
https://livewire.laravel.com/docs/loading#basic-usage

Maybe there is a directive for this.

```html
<div data-wp-navigation-id="mynavigation">
    <div data-wp-navigation-loading-id="mynavigation">
        Loading...
    </div>
 </div>
```

Or, do we export a loading state from navigation? Or maybe this is already solved.

**Comment by luisherranz** [comment]
I think it's confusing to use `navigation` in `data-wp-navigation-id` if navigation is going to be the namespace of the Navigation block (`"core/navigation"`). 

If `router` is going to be the namespace of the router (`"core/router"`), you should use something that has the router namespace in it, like: `data-wp-router-region` or `data-wp-router-region-id`.

What do you think, @darerodz?

**Comment by DAreRodz** [comment]
Hey @ajgagnon, loading states still need to be implemented, but it's already in our roadmap (see https://github.com/WordPress/gutenberg/issues/53740), so it'll be implemented soon.

Our current idea is to expose the current loading status in the global state―under the `router` namespace.

**Comment by DAreRodz** [comment]
> you should use something that has the router namespace in it, like: `data-wp-router-region` or `data-wp-router-region-id`

Agree, and I think those names are easy to understand as well, so you have my +1 👍.

**Comment by luisherranz** [comment]
Let's go with `data-wp-router-region-id` then. I think the `id` conveys the uniqueness requirement so it's a bit more explicit.

**Comment by DAreRodz** [comment]
@luisherranz, I assume we need to add the `data-wp-router-link` as well, am I right?

**Comment by luisherranz** [comment]
> @luisherranz, I assume we need to add the data-wp-router-link as well, am I right?

No, let's keep the navigate logic inside actions instead of directives for now.

**Comment by luisherranz** [comment]
I've just noticed a problem while working on `data-wp-key`: the `data-wp-router-region-id` needs to match with `data-wp-interactive`. If not, Preact recreates all the nodes from scratch (it doesn't do a diff of the DOM). This is because Preact's vDOM starts with `data-wp-interactive`, but when you do a navigation, it starts with `data-wp-router-region-id` so Preact doesn't have a previous "root" node.

Therefore, I think this should be an option of `data-wp-interactive` instead of a different directive. Something like: `data-wp-interactive='{ "region-id": "some-id" }'`. Regions inside other regions are ignored (like `data-wp-interactive`).

I also noticed that we are doing the `toVdom` twice for the initial page (once for `data-wp-interactive` and once for `data-wp-router-region-id`). If we make this change, we can save the vDOM of each `data-wp-interactive` that has a `region-id` and avoid that extra processing.

What do you think? Do you see any problem or inconvenience?

**Comment by ajgagnon** [comment]
> Hey @ajgagnon, loading states still need to be implemented, but it's already in our roadmap (see #53740), so it'll be implemented soon.
> 
> Our current idea is to expose the current loading status in the global state―under the `router` namespace.

Thanks! I did see that right after I posted 🤦🏼‍♂️  Global state makes a lot of sense!

**Comment by alexstine** [comment]
I am not following this PR for now as it looks to still be under dev but please note the accessibility concerns around using JS this way. Dynamically rendering parts of the page comes with huge implications for people who can't see changes happening if not properly communicated. With great power comes great responsibility. This is still a problem in Gutenberg today so I am super hesitant to make this available to the front-end.

Please ping me when this is ready for proper review.

Thanks.

**Comment by luisherranz** [comment]
Hi @alexstine. Yes, we're aware of the accessibility challenges, and we have plans to work on them. 

At this moment, we're only experimenting with this [here](https://github.com/WordPress/gutenberg/pull/53812) and [here](https://github.com/WordPress/gutenberg/pull/53737), so my plan is to address the accessibility in those PRs. I was planning to ping you once that initial accessibility work was done so you can review it 🙂

Once we learn more, we can find ways to include accessibility on the underlying APIs to ensure people building on top of this make things right (the Interactivity API is still experimental so nobody should be using it in production yet).

**Comment by luisherranz** [comment]
> Global state makes a lot of sense!

Hey @ajgagnon, I think we're going to add it manually first. Once we learn more, we can look into ways to make it globally accessible. Same for changing the scroll position and accessibility.

So for now, we'll use something like this (simplified):

```js
store({
  actions: {
    core: {
      query: {
        navigate: async ({ ref, context }) => {
          context.core.query.isLoading = true;
          speak(state.core.query.loadingPageMessage);
          await navigate(ref.href);
          context.core.query.isLoading = false;
          speak(state.core.query.pageLoadedMessage);
          const firstFocusableItem = getFirstFocusableItem(ref);
          firstFocusableItem.focus();
          firstFocusableItem.scrollIntoView();
        },
      },
    },
  },
});
```

File: packages/interactivity/CHANGELOG.md [hunk 1]
```diff
@@ -2,10 +2,15 @@
...
 
+### Enhancements
+
+-   Merge new server-side rendered context on client-side navigation. ([#53853](https://github.com/WordPress/gutenberg/pull/53853))
+
 ## 2.1.0 (2023-08-16)
 
 ### New Features
 
+-   Support region-based client-side navigation. ([#53733](https://github.com/WordPress/gutenberg/pull/53733))
```
Comment (luisherranz) [inline packages/interactivity/CHANGELOG.md]: For some reason, this has been moved to the 2.1.0 section when it should be in the Unreleased section.
```suggestion
### Enhancements

-   Merge new server-side rendered context on client-side navigation. ([#53853](https://github.com/WordPress/gutenberg/pull/53853))
-   Support region-based client-side navigation. ([#53733](https://github.com/WordPress/gutenberg/pull/53733))

## 2.1.0 (2023-08-16)

### New Features

```

**Comment by luisherranz** [comment]
Let's make [the `data-wp-interactive` change I mentioned previously](https://github.com/WordPress/gutenberg/pull/53733#issuecomment-1684132368) in a separate PR. I think this is good enough as it is so people can start testing the enhanced UX of the Query and Comments blocks as soon as possible.

**Review by luisherranz — COMMENTED** [review]
The tests are great, David!

I can't approve because it's my own PR, but consider them approved 🙂

**Review by DAreRodz — APPROVED** [review]

**Comment by DAreRodz** [comment]
Docs for the region-based router directives still need to be included. They'll be added in a subsequent PR.

**Pull request merged**